### PR TITLE
JAMES 2151 Github 994  Administer Sieve quotas via webadmin

### LINKED ISSUE #994: [Hacktoberfest] Administer Sieve quotas via CLI
Hello dear HocktoberFester, and thanks for considering helping us!

The goal of this issue is to contribute to https://github.com/apache/james-project (yes, an **Apache project!**) but for technical reasons, we could not open tickets on that repository...

You can gather a bit more information on what is James and why it's an awesome project [on our website](http://james.apache.org/).

-------------------------------------------------------

**Here are the more specific task details:**

Link to JIRA: https://issues.apache.org/jira/browse/JAMES-2151

James allows users to store Sieve scripts. As defined in ManageSieve RFC such storage should be limited by quotas. Hence, James allow administration of such quotas via the Command Line Client. But miss this administration aspect through *webadmin*.

In *webadmin-data* project, crete a *SieveQuota* route class directly calling *SieveRepository* to create, read and delete such quotas. Globally, and per user. You will write relevant unit tests.

You can have a look to the existing *GlobalQuotaRoutes* and related tests in *GlobalQuotaRoutesTest*.

Here is the proposed API:

```
GET /sieve/quota
150000
(Expressed in bytes)

PUT /sieve/quota
160000

DELETE /sieve/quota


GET /sieve/quota/:user
150000
(Expressed in bytes)

PUT /sieve/quota/:user
160000

DELETE /sieve/quota/:user
```
Don't hesitate to ask for help on the Gitter chat : https://gitter.im/apache/james-project

------------------------------------------------------------------------

giphy

### PR DESCRIPTION
**Pull request by sebast26**
This PR resolves #994

### CODE REVIEW

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            @ApiResponse(code = 400, message = "The body is not a positive integer."),
+            @ApiResponse(code = 500, message = "Internal server error - Something went bad on the server side.")
+    })
+    public void defineUpdateGlobalSieveQuota() {
+        service.put(ROOT_PATH, (request, response) -> {
+            try {
+                final Long requestedSize = extractRequestedQuotaSizeFromRequest(request);
+                sieveRepository.setQuota(requestedSize);
+                response.status(200);
+                return sieveRepository.getQuota();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java]: put should not return any value IMO

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/InMemorySieveQuotaRepository.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.james.sieverepository.api.exception.ScriptNotFoundException;
+import org.apache.james.sieverepository.api.exception.StorageException;
+import org.joda.time.DateTime;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemorySieveQuotaRepository implements SieveRepository {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/InMemorySieveQuotaRepository.java]: please split SieveRepository in two to create a SieveRepositoryQuota.
That way, you would not need so many empty methods in this implementation.

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/InMemorySieveQuotaRepository.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
+package org.apache.james.webadmin.routes;
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/InMemorySieveQuotaRepository.java]: license missing

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
+package org.apache.james.webadmin.routes;
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java]: license missing

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+import spark.Service;
+
+import javax.inject.Inject;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import static org.apache.james.webadmin.Constants.SEPARATOR;
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java]: We usually put `static` imports first.

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    @ApiOperation(value = "Reading global sieve quota size")
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "OK", response = Long.class),
+            @ApiResponse(code = 404, message = "Global sieve quota not set."),
+            @ApiResponse(code = 500, message = "Internal server error - Something went bad on the server side.")
+    })
+    public void defineGetGlobalSieveQuota() {
+        service.get(ROOT_PATH, (request, response) -> {
+            try {
+                final long sieveQuota = sieveRepository.getQuota();
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveQuotaRoutes.java]: You don't need `final` (same in other methods)

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/SieveQuotaRoutesTest.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
+package org.apache.james.webadmin.routes;
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/SieveQuotaRoutesTest.java]: license missing

File: server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/SieveQuotaRoutesTest.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+import org.apache.james.metrics.logger.DefaultMetricFactory;
+import org.apache.james.sieverepository.api.SieveRepository;
+import org.apache.james.webadmin.WebAdminServer;
+import org.apache.james.webadmin.WebAdminUtils;
+import org.apache.james.webadmin.utils.JsonTransformer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import static com.jayway.restassured.RestAssured.given;
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/SieveQuotaRoutesTest.java]: `static` imports first

**Comment by sebast26** [comment]
Hi @mbaechler and @aduprat I updated my PR with your suggestions.

**Comment by trantienduchn** [comment]
do we have to update `manage-webadmin.md`?

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

We did enhance your work on this pull request https://github.com/linagora/james-project/pull/1496 .

I also wanted to apologize for not giving earlier the attention your excellent work deserved. I just came around it recently and felt very sad about it being in stand-by. (next time don't hesitate to ping us. You won't be disturbing at all!)

Thank you again for your dedication toward the James project. If you need help, want to [contribute](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20newbie%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC) [documentation](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20documentation%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), new [features](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20feature%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), tests or [bugfix](https://issues.apache.org/jira/issues/?jql=project%20%3D%20JAMES%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20%3D%20easyfix%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC), please don't hesitate to reach us.

Thanks again,

Cheers,

Benoit

**Pull request closed without merging**